capsules/src/net/lowpan_fragment.rs

```diff
@@ -0,0 +1,863 @@
+            // If this fails, we received an overlapping fragment. We can simply
+            // drop the packet in this case.
+            Err(ReturnCode::FAIL)
+        } else {
+            self.bitmap
+                .map(|bitmap| bitmap.is_complete((dgram_size as usize) / 8))
+                .ok_or(ReturnCode::FAIL)
+        }
+    }
+
+    fn end_receive(&self, client: Option<&'static ReceiveClient>, result: ReturnCode) {
+        self.busy.set(false);
+        self.bitmap.map(|bitmap| bitmap.clear());
+        self.timeout_counter.set(0);
+        if client.is_some() {
+            let mut buffer = self.packet.take().unwrap();
+            client.unwrap().receive(&buffer, self.dgram_size.get(), result);
+            self.packet.replace(buffer);
+        }
+    }
+}
+
+/// struct FragState
+/// ----------------
+/// This struct tracks the global sending/receiving state, and contains the
+/// lists of RxStates and TxStates.
+pub struct FragState<'a, A: time::Alarm + 'a> {
+    pub radio: &'a Mac,
+    ctx_store: &'a ContextStore,
+    alarm: &'a A,
+
+    // Transmit state
+    tx_states: List<'a, TxState<'a>>,
+    tx_dgram_tag: Cell<u16>,
+    tx_busy: Cell<bool>,
+    tx_buf: TakeCell<'static, [u8]>,
+
+    // Receive state
+    rx_states: List<'a, RxState<'a>>,
+    rx_client: Cell<Option<&'static ReceiveClient>>,
+}
+
+// This function is called after transmitting a frame
+#[allow(unused_must_use)]
+impl<'a, A: time::Alarm> TxClient for FragState<'a, A> {
+    fn send_done(&self, buf: &'static mut [u8], acked: bool, result: ReturnCode) {
+        self.tx_buf.replace(buf);
+        if result != ReturnCode::SUCCESS {
+            self.end_packet_transmit(acked, result);
+            return;
+        }
+        self.tx_states.head().map(move |head| {
+            if head.is_transmit_done() {
+                // This must return Some if we are in the closure - in particular,
+                // tx_state == head
+                self.end_packet_transmit(acked, result);
+            } else {
+                // Otherwise, we found an error
+                // TODO
+                let tx_buf = self.tx_buf.take().unwrap();
+                let result = head.prepare_transmit_next_fragment(tx_buf, self.radio);
+                result.map_err(|(retcode, ret_buf)| {
+                    self.end_packet_transmit(acked, retcode);
+                    self.tx_buf.replace(ret_buf);
+                });
+            }
+        });
+    }
+}
+
+// This function is called after receiving a frame
+impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
+    fn receive<'b>(&self, buf: &'b [u8], header: Header<'b>, data_offset: usize, data_len: usize) {
+        // We return if retcode is not valid, as it does not make sense to issue
+        // a callback for an invalid frame reception
+        let data_offset = data_offset;
+        // TODO: Handle the case where the addresses are None/elided
+        let src_mac_addr = header.src_addr.unwrap();
```

Comment:
There are _a lot_ of `unwrap`s all over this PR. They should basically all be removed in favor of `map`, `match` or `if let Some(my_var) = my_option`. If there are places where it really really really doesn't make sense to have a case where the variable is `None`, either reconsider the type of the variable/field, or at the very least, replace it with `expect` and meaningful error message describing why it's an impossible result.

Reply:
There's a couple places where I'm not sure what the desired behavior should be. For example, in when sending a new packet and the `tx_buf` field of the `TxState` struct is `None`, then we currently panic. We don't *need* to panic here, but this case should not happen (and therefore represents a logic bug), and even if we return a failure state instead of panicking, we can never send using the same `TxState` struct (as the buffer is now permanently lost). There are 2 more analogous cases I can identify in the code. Should the code "fail fast" and panic (by calling `expect`) in these cases, or is it preferable to keep running in an inconsistent state?